tls/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+            error!(self.log, "Failed to verify cert: {e}");
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
+        })?;
+
+        info!(self.log, "Certificate chain verified successfully");
+
+        Ok(())
+    }
+
+    fn verify_signature(
+        &self,
+        message: &[u8],
+        cert: &rustls::pki_types::CertificateDer<'_>,
+        sig: &[u8],
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error>
+    {
+        // Get the public key
+        let cert = Certificate::from_der(cert).map_err(|_| {
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
+        })?;
+
+        let pubkey: [u8; 32] = match cert
+            .tbs_certificate
+            .subject_public_key_info
+            .subject_public_key
+            .as_bytes()
+        {
+            Some(pubkey) => pubkey.try_into().map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?,
+            None => {
+                return Err(rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                ))
+            }
+        };
+        let verifying_key = ed25519_dalek::VerifyingKey::from_bytes(&pubkey)
+            .map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?;
+
+        // We must hash with SHA-512 and then verify the digest
+        let mut prehashed = Sha512::new();
+        prehashed.update(message);
+
+        let signature =
+            ed25519_dalek::Signature::from_slice(sig).map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?;
+        verifying_key
+            .verify_prehashed(prehashed, Some(TLS_SIGNING_CONTEXT), &signature)
+            .map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadSignature,
+                )
+            })?;
+
+        info!(self.log, "Signature verified successfully");
+
+        Ok(HandshakeSignatureValid::assertion())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rustls::client::ResolvesClientCert;
+    use rustls::server::ResolvesServerCert;
+    use slog::Drain;
+    use std::{
+        io::ErrorKind,
+        net::{TcpListener, TcpStream},
+    };
+
+    fn logger() -> slog::Logger {
+        let decorator = slog_term::TermDecorator::new().build();
+        let drain = slog_term::FullFormat::new(decorator).build().fuse();
+        let drain = slog_async::Async::new(drain).build().fuse();
+        slog::Logger::root(drain, slog::o!("component" => "sprockets"))
+    }
```

Comment:
nit, take it or leave it: perhaps the test logger setup could be shared between the `lib.rs` tests and the tests in the other modules?

Reply:
Good idea. Done in [6b046c3](https://github.com/oxidecomputer/sprockets/pull/62/commits/6b046c3d0898f883d2f0191805c4d0b055cc89a0)